### DISCUSSION TITLE
Code repository layout and porgramming languages

### DISCUSSION THREAD

**Discussion by koh-osug** [magma/magma#14374] [Q&A]
1. Is there a detailed overview for the different project folders and their sub layout in the code repository? 
2. I see at least C++, Go and Python being used. Why at least 3 different languages and what is their purpose?
3. I see Makefiles and Bazel. Why 2 tools and what is their purpose?

**Comment by sebathomas** [comment]
1. The folder layout tends to confuse newcomers a bit (especially the second folder telling you whether something runs in the cloud or on a gateway). It's explained here in the documentation: https://magma.github.io/magma/docs/next/orc8r/architecture_modularity#modules
2. I think this is mostly for historic reasons. New components are written in Go.
3. Magma is currently migrating to a Bazel build, but the migration is not completely finished. See: https://github.com/magma/magma/issues/8338

**Reply by koh-osug** [reply]
Incredible answering speed.Thanks. 

Regarding 2: Regarding Go. What about the OAI LTE core? Isn't it in C? This is also being migrated to Go?
Regarding 3: If Go is used why not use the Go build tools? Or is their missing something?

**Reply by sebathomas** [reply]
2. You're right, MME was originally in C, and is being migrated to C++ (since this can be done without a full rewrite).
3. Actually, to be more precise, Bazel is only used in the AGW (which is mostly Python and C++ with a little Go). The Orc8r is entirely written in Go and also uses Go tooling.
